README.md
# flutter_combo_box

This package help to use custom spinner with title, title with subtitle, title with icon.

[![Build Status](https://api.cirrus-ci.com/github/flutter/plugins.svg)](https://github.com/ChrisMukasa/flutter_combo_box/blob/master)
[![Release Status](https://github.com/flutter/plugins/actions/workflows/release.yml/badge.svg)](https://github.com/ChrisMukasa/flutter_combo_box/actions/workflows/release.yml)

## Usage
[Example link](https://github.com/ChrisMukasa/flutter_combo_box/blob/master/example/example.dart)

To use this plugin, add flutter_combo_box as a [dependency in your pubspec.yaml file.](https://flutter.dev/docs/development/platform-integration/platform-channels)

### Hou to use this package
  
```yaml
dependencies:
  flutter:
    sdk: flutter
  flutter_combo_box: ^0.0.2+3 # 👈🏼 add this line
```

* Import the package to your main file
```dart
import 'package:flutter_combo_box/flutter_combo_box.dart';
```
## Combobox Tile Types
* Combobox with Title
* ComboBox with Title and subtitle
* ComboBox with Icon and title
### Title
```dart
TileTitle({String title, Color accent = Colors.blue})
```
### Title and subtitle
```dart
TileTitleSubTitle({String title, String description, Color accent = Colors.purple})
```
### Icon and title
```dart
TileIconTitle({IconData icon, String title, Color background = Colors.indigo})
```
## Example
```dart
import 'package:flutter/material.dart';
import 'package:flutter_combo_box/components/combo_box.dart';
import 'package:flutter_combo_box/flutter_combo_box.dart';

class ExampleComboBoxPage extends StatefulWidget {

  @override
  _ExampleComboBoxPageState createState() => _ExampleComboBoxPageState();
}

class Gender{
  final String? title;
  final String? subtitle;

  Gender({this.title, this.subtitle});
}

class _ExampleComboBoxPageState extends State<ExampleComboBoxPage> {
  
  var genders = ['Male', 'Female', 'Undefined'];
  var genderTitleSubTitles = [
    Gender(title: 'Male'     , subtitle: 'The male gender'),
    Gender(title: 'Female'   , subtitle: 'The female gender'),
    Gender(title: 'Undefined', subtitle: 'The undefined gender')
  ];

  late Future<List<dynamic>> datas;

  String selectedGender = '';
  var selectedGenderTitleSubTitle = Gender();

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.white,
      body: Container(
        padding: EdgeInsets.symmetric(horizontal: 24),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.center,
          children: [
            Container(
              child: ComboBoxFuture(
                label: 'Custom label', 
                hint: 'Custom hint', 
                datas: datas,
                onChanged: (item){},
              )
            ),
            SizedBox(height: 16),
            Container(
              child: Center(
                child: DropdownButtonFormField<String>(
                  decoration: InputDecoration(
                    contentPadding: EdgeInsets.symmetric(horizontal: 20, vertical: 0),
                    labelText: 'Gender',
                    hintText: 'Please select the gender here',
                    hintStyle: TextStyle(color: Colors.grey, fontSize: 16, fontWeight: FontWeight.w800),
                    alignLabelWithHint: true,
                    border: OutlineInputBorder(
                      borderRadius: BorderRadius.circular(8),
                      borderSide: BorderSide(color: Colors.black26),
                      gapPadding: 16,
                    ),
                    enabledBorder: OutlineInputBorder(
                      borderRadius: BorderRadius.circular(8),
                      borderSide: BorderSide(color: Colors.black26),
                      gapPadding: 16,
                    ),
                    focusedBorder: OutlineInputBorder(
                      borderRadius: BorderRadius.circular(8),
                      borderSide: BorderSide(color: Colors.black26),
                      gapPadding: 16,
                    ),
                  ),
                  items: genders.map((item) {
                    return DropdownMenuItem(
                      child: TileTitle(title: item, accent: Colors.red),
                      value: item,
                    );
                  }).toList(),
                  onChanged: (value) => setState(() => selectedGender = value ?? 'N/A'),
                  value: selectedGender,
                ),
              ),
            ),
            SizedBox(height: 16),
            Container(
              child: Center(
                child: DropdownButtonFormField<Gender>(
                  decoration: InputDecoration(
                    contentPadding: EdgeInsets.symmetric(horizontal: 20, vertical: 0),
                    labelText: 'Gender with Subtitle',
                    hintText: 'Please select the gender here',
                    hintStyle: TextStyle(color: Colors.grey, fontSize: 16, fontWeight: FontWeight.w800),
                    alignLabelWithHint: true,
                    border: OutlineInputBorder(
                      borderRadius: BorderRadius.circular(8),
                      borderSide: BorderSide(color: Colors.black26),
                      gapPadding: 16,
                    ),
                    enabledBorder: OutlineInputBorder(
                      borderRadius: BorderRadius.circular(8),
                      borderSide: BorderSide(color: Colors.black26),
                      gapPadding: 16,
                    ),
                    focusedBorder: OutlineInputBorder(
                      borderRadius: BorderRadius.circular(8),
                      borderSide: BorderSide(color: Colors.black26),
                      gapPadding: 16,
                    ),
                  ),
                  items: genderTitleSubTitles.map((item) {
                    return DropdownMenuItem(
                      child: TileTitleSubTitle(title: item.title, subtitle: item.subtitle, accent: Colors.red),
                      value: item,
                    );
                  }).toList(),
                  onChanged: (value) => setState(() => selectedGenderTitleSubTitle = value ?? Gender()),
                  value: selectedGenderTitleSubTitle,
                ),
              ),
            ),
          ],
        ),
      ),
    );
  }
}
```
## Screenshots

Global view 
![alt text](https://github.com/ChrisMukasa/flutter_combo_box/blob/master/example/screenshots/1.png?raw=true)
Expanded view
![alt text](https://github.com/ChrisMukasa/flutter_combo_box/blob/master/example/screenshots/2.png?raw=true)

For help getting started with Combobox package, view our 
[Gith repository](https://github.com/ChrisMukasa/flutter_combo_box)

CHANGELOG.md
## 0.0.1

* This package help to use custom spinner with title, title with subtitle, title with icon.
## 0.0.2

* Adding latest google font version number,
* Convert all static function to a StatelessWidget and StatefullWidget
* Enable Null Safety
* Adding example for ```ComboBoxTitleSubTitle```
  
## 0.0.2+1

* Correct somes bugs
  
## 0.0.2+2

* Fix somes bugs of Null Safety
  
## 0.0.2+3

* Removing ```google_fonts``` library

## 0.0.2+4

* Adding a support of Future List Data

lib/flutter_combo_box.dart
library flutter_combo_box;

// ABOUT THE DEVELOPER
//
// Contacts:
//      - Names: Chris Simba Mukasa
//      - Mail: mukasa.chris7@gmail.com
//      - Phone: +243 994 802 444
//      - Address: Av. de la mission, Himbi 2, Goma, Nord-Kivu, RDC
//
// Skills:
//      - Android Native development,
//      - Flutter developer [Android, IOS, Web and Desktop],
//      - Database Manager,
//      - ApiRestFull developer (With pure PHP),
//      - Desktop Apps developer (VB.net, C#)
//
//  Copyright (c) Aug 2021.
//  All rights reserved.
//

import 'package:flutter/material.dart';

import 'components/combo_box.dart';

class ComboBoxFuture extends StatelessWidget {
  final String label;
  final String hint;
  final String loadingHint;
  final Color accent;
  final Future<List<dynamic>> datas;
  final ValueChanged<dynamic>? onChanged;

  const ComboBoxFuture({ Key? key, required this.label, required this.hint, this.loadingHint = '', this.accent = Colors.blue, this.onChanged, required this.datas}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return FutureBuilder<List<dynamic>>(
      initialData: [],
      future: datas,
      builder: (_, snapshot) {
        if (!snapshot.hasData) {
          Text(loadingHint);
        } else {
          return DropdownButtonFormField(
            decoration: InputDecoration(
              contentPadding: EdgeInsets.symmetric(horizontal: 20, vertical: 0),
              labelText: label,
              hintText: hint,
              hintStyle: TextStyle(color: Colors.grey, fontSize: 16, fontWeight: FontWeight.w800),
              alignLabelWithHint: true,
              border: OutlineInputBorder(
                borderRadius: BorderRadius.circular(8),
                borderSide: BorderSide(color: Colors.black26),
                gapPadding: 16,
              ),
              enabledBorder: OutlineInputBorder(
                borderRadius: BorderRadius.circular(8),
                borderSide: BorderSide(color: Colors.black26),
                gapPadding: 16,
              ),
              focusedBorder: OutlineInputBorder(
                borderRadius: BorderRadius.circular(8),
                borderSide: BorderSide(color: Colors.black26),
                gapPadding: 16,
              ),
            ),
            items: snapshot.data!.map((item) {
              return DropdownMenuItem(
                child: TileTitle(title: item, accent: accent),
                value: item,
              );
            }).toList(),
            onChanged: (value) => onChanged!(value),
          );
        }
        return Text(hint);
      },
    );
  }
}

lib/components/combo_box.dart
import 'package:flutter/material.dart';

class TileTitle extends StatelessWidget {
  final String? title;
  final Color accent;

  const TileTitle({ Key? key, this.title, this.accent = Colors.blue}) : super(key: key);

  @override
  Widget build(BuildContext context) => DropdownMenuItem(
    value: title,
    child: Container(
      child: Row(
        children: <Widget>[
        Container(
          width: 20,
          height: 20,
          decoration: BoxDecoration(shape: BoxShape.circle, color: accent),
          child: Center(
            child: Text(
              title!.substring(0, 1),
              style: TextStyle(color: Colors.white, fontSize: 16, fontWeight: FontWeight.w800)
            ),
          ),
        ),
        SizedBox(width: 16),
        Text(
          title ?? 'N/A',
          style: TextStyle(color: Colors.black, fontSize: 14, fontWeight: FontWeight.w800),
          overflow: TextOverflow.fade,
          softWrap: false,
        ),
        ],
      ),
    ),
  );
}

class TileTitleSubTitle extends StatelessWidget {
  final String? title;
  final String? subtitle;
  final Color accent;

  const TileTitleSubTitle({Key? key, this.title, this.subtitle, this.accent = Colors.purple}) : super(key: key);

  @override
  Widget build(BuildContext context) => DropdownMenuItem(
    value: title,
    child: Container(
      child: Row(
        mainAxisAlignment: MainAxisAlignment.start,
        children: <Widget>[
        Container(
          width: 30,
          height: 30,
          decoration: BoxDecoration(
            shape: BoxShape.circle,
            color: accent,
          ),
          child: Center(
            child: Text(
              title!.substring(0, 1),
              style: TextStyle(color: Colors.white, fontSize: 16, fontWeight: FontWeight.w800)
            ),
          ),
        ),
        SizedBox(width: 16),
        Column(
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            Text(
              title ?? 'N/A',
              style: TextStyle(color: Colors.black, fontSize: 14, fontWeight: FontWeight.bold),
              overflow: TextOverflow.fade,
              softWrap: false,
            ),
            Text(
              subtitle ?? 'N/A',
              style: TextStyle(color: Colors.grey, fontSize: 10),
              overflow: TextOverflow.ellipsis,
              softWrap: false,
            ),
          ],
        ),
        ],
      ),
    ),
  );
}

class TileIconTitle extends StatelessWidget {
  final IconData? icon;
  final String? title;
  final Color? accent;

  const TileIconTitle({ Key? key, this.icon, this.title, this.accent }) : super(key: key);

  @override
  Widget build(BuildContext context) => DropdownMenuItem(
    value: title,
    child: Container(
      child: Row(
        children: <Widget>[
          Container(
            width: 30,
            height: 30,
            decoration: BoxDecoration(
              shape: BoxShape.circle,
              color: accent,
            ),
            child: Center(child: Icon(icon, color: Colors.white, size: 18)),
          ),
          SizedBox(width: 16),
          Text(
            title  ?? 'N/A',
            style: TextStyle(color: Colors.black, fontSize: 14, fontWeight: FontWeight.bold),
            overflow: TextOverflow.fade,
            softWrap: false,
          ),
        ],
      ),
    ),
  );
}

example/example.dart
import 'package:flutter/material.dart';
import 'package:flutter_combo_box/components/combo_box.dart';
import 'package:flutter_combo_box/flutter_combo_box.dart';

class ExampleComboBoxPage extends StatefulWidget {

  @override
  _ExampleComboBoxPageState createState() => _ExampleComboBoxPageState();
}

class Gender{
  final String? title;
  final String? subtitle;

  Gender({this.title, this.subtitle});
}

class _ExampleComboBoxPageState extends State<ExampleComboBoxPage> {
  
  var genders = ['Male', 'Female', 'Undefined'];
  var genderTitleSubTitles = [
    Gender(title: 'Male'     , subtitle: 'The male gender'),
    Gender(title: 'Female'   , subtitle: 'The female gender'),
    Gender(title: 'Undefined', subtitle: 'The undefined gender')
  ];

  late Future<List<dynamic>> datas;

  String selectedGender = '';
  var selectedGenderTitleSubTitle = Gender();

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.white,
      body: Container(
        padding: EdgeInsets.symmetric(horizontal: 24),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.center,
          children: [
            Container(
              child: ComboBoxFuture(
                label: 'Custom label', 
                hint: 'Custom hint', 
                datas: datas,
                onChanged: (item){},
              )
            ),
            SizedBox(height: 16),
            Container(
              child: Center(
                child: DropdownButtonFormField<String>(
                  decoration: InputDecoration(
                    contentPadding: EdgeInsets.symmetric(horizontal: 20, vertical: 0),
                    labelText: 'Gender',
                    hintText: 'Please select the gender here',
                    hintStyle: TextStyle(color: Colors.grey, fontSize: 16, fontWeight: FontWeight.w800),
                    alignLabelWithHint: true,
                    border: OutlineInputBorder(
                      borderRadius: BorderRadius.circular(8),
                      borderSide: BorderSide(color: Colors.black26),
                      gapPadding: 16,
                    ),
                    enabledBorder: OutlineInputBorder(
                      borderRadius: BorderRadius.circular(8),
                      borderSide: BorderSide(color: Colors.black26),
                      gapPadding: 16,
                    ),
                    focusedBorder: OutlineInputBorder(
                      borderRadius: BorderRadius.circular(8),
                      borderSide: BorderSide(color: Colors.black26),
                      gapPadding: 16,
                    ),
                  ),
                  items: genders.map((item) {
                    return DropdownMenuItem(
                      child: TileTitle(title: item, accent: Colors.red),
                      value: item,
                    );
                  }).toList(),
                  onChanged: (value) => setState(() => selectedGender = value ?? 'N/A'),
                  value: selectedGender,
                ),
              ),
            ),
            SizedBox(height: 16),
            Container(
              child: Center(
                child: DropdownButtonFormField<Gender>(
                  decoration: InputDecoration(
                    contentPadding: EdgeInsets.symmetric(horizontal: 20, vertical: 0),
                    labelText: 'Gender with Subtitle',
                    hintText: 'Please select the gender here',
                    hintStyle: TextStyle(color: Colors.grey, fontSize: 16, fontWeight: FontWeight.w800),
                    alignLabelWithHint: true,
                    border: OutlineInputBorder(
                      borderRadius: BorderRadius.circular(8),
                      borderSide: BorderSide(color: Colors.black26),
                      gapPadding: 16,
                    ),
                    enabledBorder: OutlineInputBorder(
                      borderRadius: BorderRadius.circular(8),
                      borderSide: BorderSide(color: Colors.black26),
                      gapPadding: 16,
                    ),
                    focusedBorder: OutlineInputBorder(
                      borderRadius: BorderRadius.circular(8),
                      borderSide: BorderSide(color: Colors.black26),
                      gapPadding: 16,
                    ),
                  ),
                  items: genderTitleSubTitles.map((item) {
                    return DropdownMenuItem(
                      child: TileTitleSubTitle(title: item.title, subtitle: item.subtitle, accent: Colors.red),
                      value: item,
                    );
                  }).toList(),
                  onChanged: (value) => setState(() => selectedGenderTitleSubTitle = value ?? Gender()),
                  value: selectedGenderTitleSubTitle,
                ),
              ),
            ),
          ],
        ),
      ),
    );
  }
}